Add spring-data-geode examples module

yozaner1324 opened the pull request:
The Spring Data Geode team has an example repository, so I'm incorporating it here.

mp911de commented:
Thanks for the pull request, Patrick. Can we simplify things? Having Kotlin duplicating the Java samples is a bit of an overkill. We rather should have a few dedicated Kotlin examples.

For `common-domain-models`: Please extract the common utilities (Listeners, cache writer) into a `util` project that indicates through its name what's in it. For the domain classes, we prefer a simple model that consists of one or two entities that are part of the actual example. With this approach, examples are easier to comprehend.

It would also make sense to flatten the package structure. Typically, we keep services, config, and repositories in a single package that is dedicated to one example (see [here](https://github.com/spring-projects/spring-data-examples/tree/master/jpa/example/src/main/java/example/springdata/jpa/custom) and [here](https://github.com/spring-projects/spring-data-examples/tree/master/jpa/example/src/main/java/example/springdata/jpa/caching)). We also aggregate examples as much as possible and try to reduce the number of individual subprojects. Having a lot of projects in place asks for increased maintenance. Ideally, all samples of e.g. `client-server-examples` would reside in the same project unless there's a configuration need that prevents such an layout.

Looking at the `…Service` implementations, they are basically delegates without adding value to the example. While they are best practice for real-world applications in terms of design, they add to our examples with complexity as consumers of the samples are required to step through yet another indirection.

Please also make sure to remove any `System.out`/`System.err` outputs.

After simplification of the samples, we should be able to integrate this PR.

yozaner1324 commented:
Thanks for the feedback, I'll address the things you mentioned.

yozaner1324 commented:
@mp911de I just pushed a new commit addressing some of the things you mentioned. However, I didn't do everything you requested. 

The rest of the team and I would like to keep the Kotlin versions of the examples and we'd prefer to keep the package structure as-is. Would it be possible to get these integrated without those changes?

If not, or if there is anything else that needs correcting, let me know.

yozaner1324 commented:
With my last commit, I removed the Kotlin versions, removed unnecessary services, removed all System.out/System.err outputs, made each example self-contained rather than depending on any common domain/utility classes, and flattened the package structure as much as possible.

mp911de commented on geode/advanced-examples/cascading-functions/src/main/java/example/springdata/geode/functions/cascading/Address.java:
```diff
@@ -0,0 +1,39 @@
+/*
+ *
```
Nit: License headers should look like https://github.com/spring-projects/spring-data-examples/blob/master/jpa/deferred/src/main/java/example/Application.java#L1-L15

yozaner1324 replied:
Updated the headers.

mp911de commented on geode/advanced-examples/cascading-functions/pom.xml:
```diff
@@ -0,0 +1,62 @@
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>cascading-functions</artifactId>
```
Almost every pom file contains an excessive dependency configuration with several additional dependencies exclusions. It would make sense to move the initial dependency setup into `geode/pom.xml` and adjust the pom only for the modules where it's required to minimize the cognitive load per example.

yozaner1324 replied:
Moved common dependencies to parent pom.

mp911de commented on geode/advanced-examples/cascading-functions/src/main/java/example/springdata/geode/functions/cascading/Customer.java:
```diff
@@ -0,0 +1,80 @@
+ */
+
+@Region(name = "Customers")
+public class Customer implements Serializable {
```
We're making use of Project Lombok in our examples to not pollute the code with getters/setters/equals/hashCode methods. Using Lombok for all of the domain types would help in reducing the boilerplate code to the necessary bits.

yozaner1324 replied:
Data objects now use Lombok.

mp911de commented on geode/client-server-examples/basic-operations/src/test/java/example/springdata/geode/client/basic/BasicClientTest.java:
```diff
@@ -0,0 +1,114 @@
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.NONE, classes = BasicClientApplicationConfig.class)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_CLASS)
+public class BasicClientTest extends ForkingClientServerIntegrationTestsSupport {
```
Please rename all test classes from `…Test` to `…Tests`. Right now, running `mvn test` does not execute the test classes (inclusion pattern is set to `**/*Tests.java`). We also should make sure that tests complete within a reasonable time.

yozaner1324 replied:
Renamed.

mp911de commented:
Thanks a lot, that's merged and polished now.
